Addition of packedCandidates track chi2 information to pp_on_AA miniAOD

abaty opened the pull request:
#### PR description:

This add a producer that makes two value maps which store the normalized-chi2 values of the tracks attached to objects in the PackedPFCandidate and lostTracks collections for mini-AOD in the pp_on_AA eras.  These normalized chi2 values are used in many heavy ion analyses as part of the track selection.

#### PR validation:

This was checked using the config in /afs/cern.ch/user/a/abaty/work/public/forPeople/forChi2PR, which was used for testing the heavy ion miniAOD workflow.  Both new collections (for lostTracks and PackedPFCandidates) are present and contain reasonable values.

@mandrenguyen

abaty committed:
added chi2 map for packedPFCandidates

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16803

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544, #30313
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544, #30313


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16803/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16803/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

slava77 commented:
```uint8_t normalizedChi2_``` is saved in the packed cands. Is this too coarse?

abaty commented:
> `uint8_t normalizedChi2_` is saved in the packed cands. Is this too coarse?

Yes, we found that this variable is not precise enough for our needs because it is truncated to a uint8.

slava77 commented:
> Code check has found code style and quality issues which could be resolved by applying following patch(s)
> 
>     * **code-format**:
>       https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16803/code-format.patch

this is required before further tests of this PR can be done

abaty committed:
added code style patch

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16856

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544, #30313, #30576
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544, #30313

cmsbuild commented:
A new Pull Request was created by  @abaty  for master.

It involves the following packages:

PhysicsTools/PatAlgos


@perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please review it and eventually sign? Thanks.
@rappoccio, @gouskos, @hatakeyamak, @emilbols, @peruzzim, @seemasharmafnal, @mmarionncern, @ahinzmann, @smoortga, @jdolen, @ferencek, @jdamgov, @nhanvtran, @gkasieczka, @schoef, @andrzejnovak, @clelange, @riga, @JyothsnaKomaragiri, @mverzett, @gpetruc, @mariadalfonso this is something you requested to watch as well.
@silviodonato, @dpiparo you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 99b9c449ecf9aab62531bdcd4a7c943fa3a1a3cd
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d0e872/7787/summary.html
CMSSW: CMSSW_11_2_X_2020-07-07-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d0e872/7787/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2787364
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2787313
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 154 log files, 17 edm output root files, 37 DQM output files

santocch commented:
+1

jpata commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,111 @@
+    iEvent.getByToken(trackToken_, trks);
+  }
+
+  const auto& nCand = candidates->size();
```
minor, but `const auto nCand = candidates->size()` should be fine for a simple number, this seems to be used more often in the codebase for this.

jpata commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,111 @@
+      nChi2 = candTrack->normalizedChi2();
+    }
+
+    trkChi2Map.at(i) = nChi2;
```
For quantities stored in MINIAOD, we should aim for the minimal necessary precision, as is done in e.g. https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/plugins/PATMuonSlimmer.cc#L150. You mention that uint8 was not sufficient in https://github.com/cms-sw/cmssw/pull/30566#issuecomment-654991843. What is the minimum number of significant digits that's needed for this variable when stored as a reduced-size float?

Going from the default of 23 to 12 bits for the mantissa gives an average relative error of ~1e-6, so perhaps you can reduce even more.

jpata commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,111 @@
+  }
+
+  // fill the value maps
+  std::unique_ptr<FloatMap> valueMap(new FloatMap());
```
Consider `std::make_unique`.

jpata reviewed (changes requested):
Looks good, but some algorithmic improvements can be considered. Can you provide some numbers on the increased event size and runtime?

slava77 commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,111 @@
+
+      produces<FloatMap>();
+    }
+    ~PackedCandidateTrackChi2Producer() override{};
```
empty overloads are not needed; please remove

slava77 commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,111 @@
+    static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+  private:
+    edm::EDGetTokenT<edm::View<pat::PackedCandidate>> candidateToken_;
+    edm::EDGetTokenT<edm::Association<reco::PFCandidateCollection>> candidate2PFToken_;
+    edm::EDGetTokenT<edm::Association<pat::PackedCandidateCollection>> track2LostTrackToken_;
+    edm::EDGetTokenT<edm::View<reco::Track>> trackToken_;
+    bool doLostTracks_;
```
try to keep all/most parameters const where possible;
the current impl allows for candidateToken_, doLostTracks_, trackToken_.

Also, what is the reason to read edm::View?
There is some (small) overhead to construct it. In this case it looks like the most derived class objects are consumed anyways and a polymorphic access via a View seems not justified.

slava77 commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,111 @@
+}  // namespace pat
+
+void pat::PackedCandidateTrackChi2Producer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<edm::View<pat::PackedCandidate>> candidates;
+  iEvent.getByToken(candidateToken_, candidates);
```
```suggestion
  auto const candidates =  iEvent.getByHandle(candidateToken_);
```
is a more compact and more modern way to access data in the event.
If a handle is not needed, use ```iEvent.get(someToken_)``` to get a ```cons&``` to an object

slava77 commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,111 @@
+  edm::Handle<edm::View<pat::PackedCandidate>> candidates;
+  iEvent.getByToken(candidateToken_, candidates);
+
+  edm::Handle<edm::Association<reco::PFCandidateCollection>> candidate2PF;
+  if (!doLostTracks_) {
+    iEvent.getByToken(candidate2PFToken_, candidate2PF);
```
could also be
```suggestion
  const edm::Association<reco::PFCandidateCollection>>* candidate2PF = nullptr;
  if (!doLostTracks_) {
    candidate2PF = &iEvent.get(candidate2PFToken_);
```
it can save on the extra cost of dereferencing an edm::Handle, which may be relevant here in the tight loops

slava77 commented on PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py:
```diff
@@ -134,7 +134,10 @@
-_pp_on_AA_extraCommands = ['keep *_packedCandidateMuonID_*_*']
+_pp_on_AA_extraCommands = [
+	'keep *_packedCandidateMuonID_*_*',
+	'keep floatedmValueMap_*TrackChi2_*_*'
```
wildcards should be avoided for (reco-related) products. Please spell out the producing module name in full

slava77 commented on PhysicsTools/PatAlgos/python/slimming/slimming_cff.py:
```diff
@@ -64,7 +64,11 @@
 from PhysicsTools.PatAlgos.packedCandidateMuonID_cfi import packedCandidateMuonID
-(pp_on_AA_2018 | pp_on_PbPb_run3).toReplaceWith(slimmingTask, cms.Task(slimmingTask.copy(), packedCandidateMuonID))
+from PhysicsTools.PatAlgos.packedPFCandidateTrackChi2_cfi import packedPFCandidateTrackChi2
+lostTrackChi2 = packedPFCandidateTrackChi2.clone(candidates = cms.InputTag("lostTracks"), doLostTracks = cms.bool(True))
```
please drop type specifications for all parameters which already exist.
This is a safer way to protect from parameter name mistakes and will also help in possible parameter migrations.

```suggestion
lostTrackChi2 = packedPFCandidateTrackChi2.clone(candidates = "lostTracks", doLostTracks = True)
```

abaty committed:
some changes based on code reviews

abaty committed:
changes based on code review

abaty committed:
forgot a comma

cmsbuild commented:
The code-checks are being triggered in jenkins.

abaty commented:
I have addressed the feedback provided above.

>Can you provide some numbers on the increased event size and runtime?

The average event size increases by 1.3 kb/evt, which is a 1.7% increase in the total file size.  The timing of the additional modules seems to be pretty negligible, with it taking around 10-100 microseconds for most events.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16922

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16922/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16922/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16922/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16922/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

jpata commented:
@abaty I see you've made some changes, but code checks are now failing and there is a conflict. Can you please update and take care of the code checks?

abaty committed:
added code formatting

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16956

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16956/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16956/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

abaty committed:
Merge branch 'master' into Chi2MapMaster

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/16958

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544

cmsbuild commented:
Pull request #30566 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

jpata commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 8bc215d57717ac0b7216af0abfb9ad442415fb05
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d0e872/7929/summary.html
CMSSW: CMSSW_11_2_X_2020-07-14-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d0e872/7929/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2787429
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2787378
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 154 log files, 17 edm output root files, 37 DQM output files

jpata commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,113 @@
+  if (doLostTracks_) {  //for Lost tracks we don't have references to PFCands, so we must loop over tracks and check keys...
+    for (size_t i = 0; i < trks->size(); i++) {
+      float nChi2 = 0;
+      //const auto& trk = trks->refAt(i);
```
Commented out code should be removed.

jpata commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,113 @@
+
+  if (doLostTracks_) {  //for Lost tracks we don't have references to PFCands, so we must loop over tracks and check keys...
+    for (size_t i = 0; i < trks->size(); i++) {
+      float nChi2 = 0;
```
remove the `float nChi2` allocation here and just move it directly to line 73, it is not used outside of the `doLostTracks_`.

jpata commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,113 @@
+  } else {  //for the regular PackedPFCands we have direct references...
+    for (size_t i = 0; i < nCand; i++) {
+      const auto& cand = pat::PackedCandidateRef(candidates, i);
+      float nChi2 = 0;
```
as above, move this directly to line 87: `const float nChi2 = candTrack->normalizedChi2();`

jpata commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,113 @@
+      const auto& candTrack = (*candidate2PF)[cand]->trackRef();
+      nChi2 = candTrack->normalizedChi2();
+
+      trkChi2Map.at(i) = MiniFloatConverter::reduceMantissaToNbitsRounding<12>(nChi2);
```
the rounding constant `12` should not be duplicated in a hardcoded way, rather make it a class constant that is propagated to both line 89 and 74. Did you check if you can use an even smaller rounding precision?

jpata commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,113 @@
+
+void pat::PackedCandidateTrackChi2Producer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<pat::PackedCandidateCollection> candidates;
+  iEvent.getByToken(candidateToken_, candidates);
```
Please check Slava's comment https://github.com/cms-sw/cmssw/pull/30566#discussion_r452809205, can you implement that here?

jpata reviewed (commented):
Thanks for the updates @abaty. I added a few more code comments, please have a look. I think it looks good, generally.

abaty committed:
reduced chi2 precision futher

cmsbuild commented:
The code-checks are being triggered in jenkins.

abaty commented:
@jpata I have implemented the remaining feedback.

>Did you check if you can use an even smaller rounding precision?

Regarding this, we have moved to using 8 bits in the mantissa. I believe this should give a relative precision of a bit under 1% for values around 1 (which the normalized chi2 distribution peaks at), which is good enough for our needs.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/17056

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544

cmsbuild commented:
Pull request #30566 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

jpata commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 0c885df71252cf2ee8af8e510f977136bbe82407
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d0e872/7999/summary.html
CMSSW: CMSSW_11_2_X_2020-07-15-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

jpata commented on PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc:
```diff
@@ -0,0 +1,110 @@
+      if (cand->charge() == 0 || !cand->hasTrackDetails())
+        continue;
+
+      const auto& candTrack = (*candidate2PF)[cand]->trackRef();
```
check for isNonnull here.

jpata commented:
@abaty I tested this PR with your config `/afs/cern.ch/user/a/abaty/work/public/forPeople/forChi2PR/reMiniAOD_DATA_PAT.py` on 500 events, merged against `CMSSW_11_2_DEVEL_X_2020-07-15-2300`

At some point, it crashes with:
```
----- Begin Fatal Exception 16-Jul-2020 10:48:35 CEST-----------------------
Begin processing the 476th record. Run 327219, Event 359191451, LumiSection 709 on stream 7 at 16-Jul-2020 10:48:35.823 CEST
Begin processing the 477th record. Run 327219, Event 359372315, LumiSection 709 on stream 6 at 16-Jul-2020 10:48:35.823 CEST
An exception of category 'InvalidReference' occurred while
   [0] Processing  Event run: 327219 lumi: 709 event: 358987571 stream: 2
   [1] Running path 'MINIAODoutput_step'
   [2] Prefetching for module PoolOutputModule/'MINIAODoutput'
   [3] Calling method for module PackedCandidateTrackChi2Producer/'packedPFCandidateTrackChi2'
Exception Message:
BadRefCore RefCore: Request to resolve a null or invalid reference to a product of type 'std::vector<reco::Track>' has been detected.
Please modify the calling code to test validity before dereferencing.
----- End Fatal Exception -------------------------------------------------
```

Now, this is caused by the `candTrack` here: https://github.com/cms-sw/cmssw/pull/30566/files#r455695038
Maybe this is rare enough not to be caught by the jenkins test, but seems like it could be an issue.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d0e872/7999/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2612943
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2612894
 - DQMHistoTests: Total skipped: 48
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 146 log files, 17 edm output root files, 35 DQM output files

abaty committed:
fixed bug with null pointer refernece

cmsbuild commented:
The code-checks are being triggered in jenkins.

abaty commented:
@jpata I was able to reproduce this bug, and it should now be fixed.  I was testing on 300 events previously.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/17095

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544

cmsbuild commented:
Pull request #30566 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: b133b57bc05d8bbed8e5e0210971b430c4277ede
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d0e872/8032/summary.html
CMSSW: CMSSW_11_2_X_2020-07-16-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d0e872/8032/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2612943
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2612894
 - DQMHistoTests: Total skipped: 48
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 146 log files, 17 edm output root files, 35 DQM output files

jpata commented:
- technical tests pass
- MINIAOD file size increases between 1.7-3% for heavy ion workflows, which is not negligible, but should be fine for a limited production
- runtime increase is reported to be negligible
- outputs are filled with meaningful values

From what I see, should be fine to approve from reco. @slava77 @perrotta what do you think?

perrotta commented:
> * technical tests pass
> * MINIAOD file size increases between 1.7-3% for heavy ion workflows, which is not negligible, but should be fine for a limited production
> * runtime increase is reported to be negligible
> * outputs are filled with meaningful values
> 
> From what I see, should be fine to approve from reco. @slava77 @perrotta what do you think?

Well, HI run3 will not be a limited production. But the alternative is to stick on AOD and not move to miniAOD for HI analyses, and therefore I'd say that the size increase can be deemed acceptable.

For the reco signature I'd let @slava77 to comment, since he also participated to the review

slava77 commented on PhysicsTools/PatAlgos/python/slimming/slimming_cff.py:
```diff
@@ -67,7 +67,11 @@
-(pp_on_AA_2018 | pp_on_PbPb_run3).toReplaceWith(slimmingTask, cms.Task(slimmingTask.copy(), packedCandidateMuonID))
+from PhysicsTools.PatAlgos.packedPFCandidateTrackChi2_cfi import packedPFCandidateTrackChi2
+lostTrackChi2 = packedPFCandidateTrackChi2.clone(candidates = "lostTracks", doLostTracks = True)
+(pp_on_AA_2018 | pp_on_PbPb_run3).toReplaceWith(slimmingTask, cms.Task(slimmingTask.copy(), packedCandidateMuonID, packedPFCandidateTrackChi2, lostTrackChi2))
+
+
```
```suggestion
(pp_on_AA_2018 | pp_on_PbPb_run3).toReplaceWith(slimmingTask, 
                                                cms.Task(slimmingTask.copy(), packedCandidateMuonID,
                                                         packedPFCandidateTrackChi2, lostTrackChi2))
```
not required, but something to make the update a bit more readable

slava77 commented:
> For the reco signature I'd let @slava77 to comment, since he also participated to the review

it's OK with me
even though there is a minor thing in https://github.com/cms-sw/cmssw/pull/30566#pullrequestreview-450636149

abaty committed:
small python formatting change for pp_on_AA

cmsbuild commented:
The code-checks are being triggered in jenkins.

abaty commented:
>even though there is a minor thing in #30566 (review)

I made a small commit addressing this comment from @slava77 .

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30566/17151

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544, #30810
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544, #30810

cmsbuild commented:
Pull request #30566 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 0cf27e88ebcd952ac3e484c2fa5e808fd5cf527f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d0e872/8095/summary.html
CMSSW: CMSSW_11_2_X_2020-07-17-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d0e872/8095/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2525996
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2525948
 - DQMHistoTests: Total skipped: 47
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 144 log files, 17 edm output root files, 34 DQM output files

jpata commented:
+1

perrotta commented:
> +1 (see motivation here: [#30566 (comment)](https://github.com/cms-sw/cmssw/pull/30566#issuecomment-660024083))

@jpata reco approval was added and removed at once by bot. I suspect it is because "+1" and the comment are written on the same line. Maybe you can try with two separate lines and see whether it is accepted as such...

jpata commented:
> > +1 (see motivation here: [#30566 (comment)](https://github.com/cms-sw/cmssw/pull/30566#issuecomment-660024083))
> 
> @jpata reco approval was added and removed at once by bot. I suspect it is because "+1" and the comment are written on the same line. Maybe you can try with two separate lines and see whether it is accepted as such...

Oops, thanks for the heads up.

santocch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
